<?xml version='1.0'?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="ddrobot">
 <xacro:property name="pi2" value="1.57079632679" />
 <xacro:property name="body_length" value="2" />
 <xacro:property name="sled_mass" value="50" />
 <xacro:property name="sled_x" value="1.5" />
 <xacro:property name="sled_y" value="1" />
 <xacro:property name="sled_z" value="0.25" />

  <xacro:macro name="wheel_macro" params="xyz name model">
    <link name="${name}">
     <inertial>
      <mass value="7.5"/>
      <origin rpy="0 0 0" xyz="${xyz}"/>
      <inertia ixx="0.49" ixy="0" ixz="0" iyy="0.37" iyz="0" izz="0.37"/>
     </inertial>
     <collision>
       <origin xyz="${xyz}" rpy="0 1.57079632679 0"/>
       <laser_retro>0</laser_retro>
       <max_contacts>10</max_contacts>
       <pose >0 0 0 0 0 0</pose>
       <geometry>
         <cylinder length="0.45" radius="0.28"/>
       </geometry>
     </collision>
     <visual name="visual">
      <origin rpy="0 0 ${pi2}" xyz="${xyz}"/>
      <geometry>
       <mesh filename="package://ros_robotics/meshes/${model}" scale="1 1 1"/>
      </geometry>
      <material name="silver">
       <color rgba="0.75 0.75 0.75 1"/>
      </material>
     </visual>
    </link>
  <xacro:wheel_bumper_macro reference="${name}" />
  </xacro:macro>


 <xacro:macro name="wheel_bumper_macro" params="reference">
 <gazebo reference="${reference}">
  <fdir1>1 0 0</fdir1>
   <!-- contact sensor -->
   <sensor type="contact" name="${reference}_contact_sensor">
     <update_rate>50.0</update_rate>
     <always_on>true</always_on>
     <contact>
       <collision>${reference}_collision</collision>
     </contact>
     <plugin name="lside_bumper_plugin" filename="libgazebo_ros_bumper.so">
       <bumperTopicName>bumper_${reference}</bumperTopicName>
       <frameName>World</frameName>
  </plugin>
   </sensor>
 </gazebo>
 </xacro:macro>






 <xacro:macro name="trans_macro" params="name">
  <transmission name="simple_trans">
   <type>transmission_interface/SimpleTransmission</type>
   <joint name="${name}">
     <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
   </joint>
   <actuator name="motor_${name}">
     <mechanicalReduction>1</mechanicalReduction>
   </actuator>
  </transmission>
 </xacro:macro>

 <xacro:macro name="pivot_joint_macro" params="xyz name parent child">
  <joint name="${name}" type="revolute">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="0  0  1"/>
   <!--limit effort="-1" lower="-0.18" upper="0.18" velocity="-1"/-->
   <limit effort="-1" lower="-0.01" upper="0.01" velocity="-1"/>
  </joint>
  <xacro:trans_macro name="${name}"/>
 </xacro:macro>

 <!--joint name="joint_pivot" type="revolute">
  <parent link="base_link"/>
  <child link="front_axle"/>
  <origin rpy="0    0   -1.57079632679" xyz="0  -0.55  0.03459"/>
  <axis xyz="0  0  1"/>
  <limit effort="-1" lower="-0.157" upper="0.157" velocity="-1"/>
 </joint-->

<xacro:macro name="joint_wheel_macro" params="xyz name parent child">
 <joint name="${name}" type="revolute">
  <parent link="${parent}"/>
  <child link="${child}"/>
  <origin rpy="0    0    1.57079632679" xyz="${xyz}"/>
  <axis xyz="1  0  0"/>
  <limit effort="200" lower="-1e+16" upper="1e+16" velocity="100"/>
  <dynamics damping="0" friction="0"/>
 </joint>
 <xacro:trans_macro name="${name}"/>
</xacro:macro>

<xacro:macro name="axle_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <mass value="15"/>
   <origin rpy="0  0  0" xyz="0  0  0"/>
   <inertia ixx="0.166667" ixy="0" ixz="0" iyy="0.166667" iyz="0" izz="0.166667"/>
  </inertial>
  <collision name="collision">
   <origin rpy="0  0  0" xyz="0    0   0"/>
   <geometry>
    <mesh filename="package://ros_robotics/meshes/Front_Axle.STL" scale="1.0 1.0 1.0"/>
   </geometry>
  </collision>
  <visual name="visual">
   <origin rpy="0  0  0" xyz="0 0 0"/>
   <geometry>
    <mesh filename="package://ros_robotics/meshes/Front_Axle.STL" scale="1 1 1"/>
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
 </link>
</xacro:macro>

<xacro:macro name="body_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <origin xyz="${xyz}"/>
   <mass value="40"/>
   <inertia ixx="2" ixy="0" ixz="0" iyy="4" iyz="0" izz="6"/>
  </inertial>
  <collision name="collision">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <box size="2 0.5 0.5" />
   </geometry>
  </collision>
  <visual name="visual">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <box size="2 0.5 0.5" />
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
  </link>
 </xacro:macro>

<xacro:macro name="sled_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <origin xyz="${xyz}"/>
   <mass value="${sled_mass}"/>
   <inertia ixx="${(1/12)*sled_mass*(pow(sled_z,2) + pow(sled_y,2))}" ixy="0" ixz="0" iyy="${(1/12)*sled_mass*(pow(sled_z,2) + pow(sled_x,2))}" iyz="0" izz="${(1/12)*sled_mass*(pow(sled_x,2) + pow(sled_y,2))}"/>
  </inertial>
  <collision>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <box size="${sled_x} ${sled_y} ${sled_z}" />
   </geometry>
  </collision>
  <visual name="visual">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <box size="${sled_x} ${sled_y} ${sled_z}" />
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
  </link>
  <xacro:wheel_bumper_macro reference="${name}" />
 </xacro:macro>

 <xacro:macro name="fixed_joint_macro" params="xyz name parent child">
  <joint name="${name}" type="prismatic">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="0 0 1"/>
   <limit effort="-1" lower="-0.01" upper="0.01" velocity="-1"/>
  </joint>
 </xacro:macro>

<xacro:macro name="dummy_link_macro" params="xyz name mass">
 <link name="${name}">
  <inertial>
   <mass value="${0.001 + mass}"/>
   <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
  </inertial>
  <visual name="visual">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <sphere radius="0.125"/>
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
  </link>
 </xacro:macro>

<xacro:macro name="body_joint_macro" params="xyz axis name parent child">
  <joint name="${name}" type="revolute">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="${axis}"/>
   <!--limit effort="-1" lower="-0.001" upper="0.001" velocity="-1"/-->
   <limit effort="-1" lower="-3.14" upper="3.14" velocity="-1"/>
  </joint>
  <xacro:trans_macro name="${name}"/>
</xacro:macro>


 <link name="base_link">
  <inertial>
   <mass value="40"/>
   <origin rpy="${pi2} 0 ${pi2}" xyz="0 0 0"/>
   <inertia ixx="2" ixy="0" ixz="0" iyy="4" iyz="0" izz="6"/>
  </inertial>
  <collision name="collision">
   <origin rpy="${pi2} 0 ${pi2}" xyz="0 0 0"/>
   <geometry>
    <mesh filename="package://ros_robotics/meshes/body.STL" scale="1.0 1.0 1.0"/>
   </geometry>
  </collision>
  <visual name="visual">
   <origin rpy="${pi2} 0 ${pi2}" xyz="0 0 0"/>
   <geometry>
    <mesh filename="package://ros_robotics/meshes/body.STL" scale="1 1 1"/>
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
 </link>

<xacro:body_joint_macro name="body_joint_1_x" parent="base_link" child="dummy_link_1_1" xyz="-0.5 0 0" axis="1 0 0"/>
<xacro:dummy_link_macro name="dummy_link_1_1" xyz="0 0 0" mass="0"/>
<xacro:body_joint_macro name="body_joint_1_y" parent="dummy_link_1_1" child="dummy_link_1_2" xyz="0 0 0" axis="0 1 0"/>
<xacro:dummy_link_macro name="dummy_link_1_2" xyz="0 0 0" mass="0"/>
<xacro:body_joint_macro name="body_joint_1_z" parent="dummy_link_1_2" child="robot_anchor" xyz="0 0 0" axis="0 0 1"/>


<xacro:dummy_link_macro name="robot_anchor" xyz="0 0 0" mass="15"/>




<joint name="robot_to_sled" type="prismatic">
   <parent link="robot_anchor"/>
   <child link="dummy_link_2_2"/>
   <origin rpy="0 0 0" xyz="-1.5 0 0"/>
   <axis xyz="1 0 0"/>
   <limit effort="-1" lower="-1000" upper="1" velocity="-1"/>
</joint>



<xacro:dummy_link_macro name="dummy_link_2_2" xyz="0 0 0" mass="0"/>
<xacro:body_joint_macro name="body_joint_2_y" parent="dummy_link_2_2" child="dummy_link_2_3" xyz="0 0 0" axis="0 1 0"/>
<xacro:dummy_link_macro name="dummy_link_2_3" xyz="0 0 0" mass="0"/>
<xacro:body_joint_macro name="body_joint_2_z" parent="dummy_link_2_3" child="sled_anchor" xyz="0 0 0" axis="0 0 1"/>


<xacro:dummy_link_macro name="sled_anchor" xyz="0 0 0" mass="0"/>
<xacro:fixed_joint_macro name="fixed_joint_2" parent="sled_anchor" child="sled" xyz="${-sled_x/2} 0 0"/>
<xacro:sled_macro name="sled" xyz="0 0 0"/>



<xacro:joint_wheel_macro xyz="-0.5075  0.57526 -0.063" name="joint_back_left" parent="base_link" child="back_left"/>
<xacro:joint_wheel_macro xyz="-0.5075 -0.57526 -0.063" name="joint_back_right" parent="base_link" child="back_right"/>
<xacro:wheel_macro  xyz="0 0 0" name="back_left" model="frostyboy_wheel.STL"/>
<xacro:wheel_macro  xyz="0 0 0" name="back_right" model="frostyboy_wheel.STL"/>

<xacro:pivot_joint_macro name="joint_pivot" parent="base_link" child="front_axle" xyz="0.55  0  -0.055"/>
<xacro:axle_macro name="front_axle" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 -0.01" name="joint_front_left" parent="front_axle" child="front_left"/>
<xacro:joint_wheel_macro xyz="0 -0.57526 -0.01" name="joint_front_right" parent="front_axle" child="front_right"/>
<xacro:wheel_macro  xyz="0 0 0" name="front_left" model="frostyboy_wheel.STL"/>
<xacro:wheel_macro  xyz="0 0 0" name="front_right" model="frostyboy_wheel.STL"/>





<gazebo>
  <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
   <robotNamespace>/dd_robot</robotNamespace>
   <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
   <legacyModeNS>true</legacyModeNS>
  </plugin>
</gazebo>
</robot>
